src/future/flatten_sink.rs

```diff
@@ -0,0 +1,91 @@
+use std::mem;
+
+use {Async, AsyncSink, StartSend, Poll, Future, Sink};
+
+
+enum State<F: Future>
+    where F::Item: Sink
+{
+    Future(F),
+    Sink(F::Item),
+    Error(<F::Item as Sink>::SinkError),
+    Void,
+}
+
+/// Future for the `into_sink` combinator, flattening a future-of-a-sink to
+/// get just the result of the final sink instead
+///
+/// This is created by the `Future::into_sink` method.
+pub struct FlattenSink<F: Future>
+    where F::Item: Sink
+{
+    state: State<F>,
+}
+
+pub fn new<F>(future: F) -> FlattenSink<F>
+    where F: Future,
+          F::Item: Sink
+{
+    FlattenSink {
+        state: State::Future(future),
+    }
+}
+
+impl<F: Future> Sink for FlattenSink<F>
+    where F::Item: Sink,
+          <F::Item as Sink>::SinkError: From<F::Error>,
+{
+    type SinkItem = <F::Item as Sink>::SinkItem;
+    type SinkError = <F::Item as Sink>::SinkError;
+    fn start_send(&mut self, item: Self::SinkItem)
+        -> StartSend<Self::SinkItem, Self::SinkError>
+    {
+        let (res, state) = match mem::replace(&mut self.state, State::Void) {
+            State::Future(mut fut) => {
+                match fut.poll() {
+                    Ok(Async::Ready(mut sink)) => {
+                        (sink.start_send(item)?, State::Sink(sink))
+                    }
+                    Ok(Async::NotReady) => {
+                        (AsyncSink::NotReady(item), State::Future(fut))
+                    }
+                    Err(e) => {
+                        (AsyncSink::NotReady(item), State::Error(e.into()))
+                    }
+                }
+            }
+            State::Sink(mut sink) => {
+                (sink.start_send(item)?, State::Sink(sink))
+            }
+            State::Void => {
+                panic!("Called FlattenSink::start_send after error");
+            }
+            State::Error(e) => (AsyncSink::NotReady(item), State::Error(e)),
```

Comment:
Yeah that use case makes sense, but relies on `poll_complete` eventually getting called. We've [already got code](https://github.com/alexcrichton/futures-rs/blob/81273d39d09746729a390e0d3ba39cfc097d840f/src/stream/forward.rs#L69-L73), though, which only calls `start_send` instead of `poll_complete`. Additionally the [current documentation on `start_send`](https://github.com/alexcrichton/futures-rs/blob/81273d39d09746729a390e0d3ba39cfc097d840f/src/sink/mod.rs#L114-L119) may indicate that this should propagate an error?

Now this does mean that the method may not be appropriate for the use case you're thinking of. In general though I think that makes sense, though? If you've got a pooled connection presumably you'd want some cloneable request to send to the pool, because even if the pool *didn't* hit an error in the future, but it hit an error in the sink, you'd want to retry the request, right?